src/asset/asset-list-loader.js

```diff
@@ -0,0 +1,215 @@
+        }
+    }
+
+    /**
+     * Removes all references to this asset list loader.
+     */
+    destroy() {
+        // remove any outstanding listeners
+        const self = this;
+
+        this._registry.off("load", this._onLoad);
+        this._registry.off("error", this._onError);
+
+        this._waitingAssets.forEach(function (id) {
+            self._registry.off("add:" + id, this._onAddAsset);
+        });
+
+        this.off("progress");
+        this.off("load");
+    }
+
+    /**
+     * Start loading asset list, call done() when all assets have loaded or failed to load.
+     *
+     * @param {Function} done - Callback called when all assets in the list are loaded. Passed (err, failed) where err is the undefined if no errors are encountered and failed contains a list of assets that failed to load.
+     * @param {object} [scope] - Scope to use when calling callback.
+     */
+    load(done, scope) {
+        const l = this._assets.length;
+        let asset;
+
+        this._count = 0;
+        this._failed = [];
+        this._callback = done;
+        this._scope = scope;
+        this._loaded = false;
+
+        this._registry.on("load", this._onLoad, this);
+        this._registry.on("error", this._onError, this);
+
+        let loadingAssets = false;
+        for (let i = 0; i < l; i++) {
+            asset = this._assets[i];
+
+            // Track assets that are not loaded or are currently loading
+            // as some assets may be loading by this call
+            if (!asset.loaded) {
+                loadingAssets = true;
+                this._loadingAssets.add(asset);
+                this._registry.add(asset);
+            }
+        }
+        this._loadingAssets.forEach((asset) => {
+            this._registry.load(asset);
+        });
+        if (!loadingAssets && this._waitingAssets.length === 0) {
+            this.fire("load", this._assets);
+        }
+    }
+
+    /**
+     * Sets a callback which will be called when all assets in the list have been loaded or failed to load.
+     *
+     * @param {Function} done - Callback called when all assets in the list are loaded.
+     * @param {object} [scope] - Scope to use when calling callback.
+     */
+    ready(done, scope = this) {
+        if (this._loaded) {
+            done.call(scope, this._assets);
+        } else {
+            this.once("load", function (assets) {
+                done.call(scope, assets);
+            });
+        }
+    }
+
+    // called when all assets are loaded
+    _loadingComplete() {
+        if (this._loaded) return;
+        this._loaded = true;
+        this._registry.off("load", this._onLoad, this);
+        this._registry.off("error", this._onError, this);
+
+        if (this._failed && this._failed.length) {
+            if (this._callback) {
+                this._callback.call(this._scope, "Failed to load some assets", this._failed);
+            }
+            this.fire("error", this._failed);
+        } else {
```

Comment:
The load event isn't called when it's failed but the load functions callback is invoked directly with the failed assets. This can be used to retry the load. I've added in a test case for this.

Reply:
That sounds good .. I added a comment to provide an example on how to use this. Should we provide some maxRetries parameter or is this somehow already available for assets loaded this way?

Reply:
The AssetListLoader just acts as a wrapper for Asset loading, so any parameters should be set on the assets themselves before they're passed into the list loader.